File: harness/fuzz/c/tmr_fuzz_harness.c
#include <stdint.h>

#include "fuzz_common.h"
#include "../../common/harness_abi.h"
#include "../../../c/tmr/tmr.h"

_Static_assert((int)TMR_OK == HARNESS_STATUS_OK,
    "tmr_status_t::TMR_OK must match HARNESS_STATUS_OK");
_Static_assert((int)TMR_ERR_NO_MAJORITY == HARNESS_STATUS_NO_MAJORITY,
    "tmr_status_t::TMR_ERR_NO_MAJORITY must match HARNESS_STATUS_NO_MAJORITY");

volatile tmr_int_t harness_fuzz_tmr_state;

static uint32_t sample_value(uint64_t *rng) {
    return 0x5a5a0000u ^ harness_random_u32(rng);
}

void harness_main(void) {
    uint64_t rng = harness_seed_state();
    const uint32_t expected = sample_value(&rng);
    int value = 0;
    tmr_status_t status;

    harness_expected = expected;
    harness_fuzz_tmr_state = tmr_int_init((int)expected);

    harness_open_fault_window();
    status = tmr_int_read((tmr_int_t *)&harness_fuzz_tmr_state, &value);
    harness_close_fault_window();

    if (status == TMR_OK) {
        harness_output = (uint32_t)value;
        if (harness_fuzz_tmr_state.fault_count != 0u) {
            harness_detected = 1u;
            if (harness_output == harness_expected) {
                harness_corrected = 1u;
            }
        }
    } else {
        harness_detected = 1u;
        harness_safe_state = 1u;
        harness_error_code = HARNESS_STATUS_NO_MAJORITY;
    }

    harness_finish();
}

File: harness/fuzz/zig/tmr_fuzz_harness.zig
const fuzz = @import("fuzz_abi.zig");
const tmr = @import("tmr");

const TmrU32 = tmr.Tmr(u32);

export var harness_fuzz_tmr_state: TmrU32 = undefined;

fn sampleValue(rng: *u64) u32 {
    return 0x5a5a0000 ^ fuzz.randomU32(rng);
}

export fn harness_main() callconv(.c) noreturn {
    var rng = fuzz.seedState();
    const expected = sampleValue(&rng);

    fuzz.store32(&fuzz.harness_expected, expected);
    harness_fuzz_tmr_state = TmrU32.init(expected);

    fuzz.openFaultWindow();
    // never_inline keeps loads of harness_fuzz_tmr_state from being hoisted across the window.
    const read_result = @call(.never_inline, TmrU32.read, .{&harness_fuzz_tmr_state});
    fuzz.closeFaultWindow();

    if (read_result) |value| {
        fuzz.store32(&fuzz.harness_output, value);
        if (harness_fuzz_tmr_state.fault_count != 0) {
            fuzz.store32(&fuzz.harness_detected, 1);
            if (value == expected) {
                fuzz.store32(&fuzz.harness_corrected, 1);
            }
        }
    } else |err| switch (err) {
        tmr.TmrError.NoMajority => {
            fuzz.store32(&fuzz.harness_detected, 1);
            fuzz.store32(&fuzz.harness_safe_state, 1);
            fuzz.store32(&fuzz.harness_error_code, 1);
        },
    }

    fuzz.complete();
}

pub const panic = fuzz.panic;
